### PR TITLE
Add prioritization for compaction flow

### PR DESCRIPTION
**Pull request by yukuai518**

### CODE REVIEW

File: gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java [hunk 1]
```diff
@@ -177,6 +209,42 @@ public void run () {
...
+      throw new RuntimeException("Cannot initialize allocator");
+    }
+  }
+
+  private List<Dataset> prioritize (List<Dataset> datasets, State state) {
+    double maxPool = state.getPropAsDouble(MRCompactor.COMPACTION_DATASETS_MAX_COUNT, MRCompactor.DEFUALT_COMPACTION_DATASETS_MAX_COUNT);
+    ResourcePool pool = ResourcePool.builder().maxResource(SimpleDatasetRequest.SIMPLE_DATASET_COUNT_DIMENSION, maxPool).build();
+
+    List<Dataset> newList = new ArrayList<>();
+    AllocatedRequestsIterator<SimpleDatasetRequest> allocatedRequests = this.allocator.allocateRequests(Iterators.transform(datasets.iterator(), new Function<Dataset, Requestor<SimpleDatasetRequest>>() {
```
Comment (ibuenros) [inline gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java]: With java 8 you can simplify:

`Lists.newArrayList(this.allocator.allocateRequests(datasets.stream().map(SimpleDatasetRequestor::new).iterator(), pool))`

File: gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java [hunk 2]
```diff
@@ -177,6 +209,42 @@ public void run () {
     }
   }
 
+  private void initAllocator (State state) {
+    try{
+      Comparator<SimpleDatasetRequest> prioritizer = new SimpleDatasetHierarchicalPrioritizer(state);
```
Comment (ibuenros) [inline gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java]: Why does this not support specifying custom prioritizer?

File: gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java [hunk 3]
```diff
@@ -177,6 +209,42 @@ public void run () {
...
+
+      RequestAllocatorConfig.Builder<SimpleDatasetRequest> configBuilder =
+          RequestAllocatorConfig.builder(new SimpleDatasetRequest.SimpleDatasetRequestEstimator()).allowParallelization(1)
```
Comment (ibuenros) [inline gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java]: Why can't user specify custom request estimator?

File: gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import gobblin.dataset.Dataset;
+import gobblin.util.request_allocation.Request;
+import gobblin.util.request_allocation.Requestor;
+import gobblin.util.request_allocation.ResourceEstimator;
+import gobblin.util.request_allocation.ResourcePool;
+import gobblin.util.request_allocation.ResourceRequirement;
+
+
+@AllArgsConstructor
+public class SimpleDatasetRequest implements Request<SimpleDatasetRequest> {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequest.java]: Please add javadoc.

File: gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequestor.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import gobblin.dataset.Dataset;
+import gobblin.util.request_allocation.PushDownRequestor;
+
+
+@AllArgsConstructor
+public class SimpleDatasetRequestor implements PushDownRequestor<SimpleDatasetRequest> {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequestor.java]: Please add javadoc.

Comment (yukuai518) [inline gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequestor.java]: Done

File: gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequest.java (hunk 4 above)
Comment (yukuai518) [inline gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequest.java]: Done

File: gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java (hunk 1 above)
Comment (yukuai518) [inline gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java]: Done

File: gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java (hunk 3 above)
Comment (yukuai518) [inline gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java]: Done

File: gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java (hunk 2 above)
Comment (yukuai518) [inline gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java]: Done

File: gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetHierarchicalPrioritizer.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ * {@link gobblin.util.request_allocation.HierarchicalAllocator} will use {@link TierComparator} from this class
+ * to shuffle {@link SimpleDatasetRequestor}s so that high priority tiers will appear in front of low priority tiers.
+ *
+ * Usage:
+ * {@link #TIER_KEY}.<tier-number>=<whitelist-blacklist-pattern>
+ * Example:
+ * {@link #TIER_KEY}.0 = pattern_0
+ * {@link #TIER_KEY}.1 = pattern_1
+ */
+@Alias("SimpleTiers")
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetHierarchicalPrioritizer.java]: Rename to `TieredDatasets`?

File: gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  public static final String SIMPLE_DATASET_COUNT_DIMENSION = "count";
+  @Getter
+  Dataset dataset;
+  SimpleDatasetRequestor requestor;
+  @Override
+  public Requestor<SimpleDatasetRequest> getRequestor() {
+    return requestor;
+  }
+
+  @Alias("SimpleEstimator")
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/dataset/SimpleDatasetRequest.java]: Either use a less generic alias or don't use an alias at all. It is perfectly ok to require the fully qualified class name to be used.

**Pull request merged**